File: Linkedin-Intership-Interview/R1Q2.java
/* merging a sorted array*/

import java.util.*;
import java.lang.*;
import java.io.*;

/* Name of the class has to be "Main" only if the class is public. */
class Ideone
{	
	public String[] merge(String[] array1, String[] array2) { //O(sum of length of two string)
		int m = array1.length;
		int n = array2.length;
		if(m == 0)	//no second array to merge
			return array2;
		if(n == 0) 	//no first array to merge
			return array1;
		String[] resultArray = new String[m+n];
		int i = 0,j = 0,k=0;
		while((i < m) && (j < n)) { //since the array is already sorted we are using a single while loop
			if(array1[i].compareTo(array2[j]) < 0)
			{
				resultArray[k] = array1[i];
				i++;k++;
			}
			else{
				resultArray[k] = array2[j];
				j++;k++;
			}	
		}
		while(i<m){	//store the remaining element in first array
			resultArray[k] = array1[i];
			i++;k++;
		}
		while(j<n){	//store the remaining element in first array
			resultArray[k] = array2[j];
			j++;k++;
		}
		return resultArray;
	}
	public static void main (String[] args) throws java.lang.Exception
	{
		Ideone program = new Ideone();
		String[] array1 = {"good", "person"};
    	String[] array2 = {"he", "is"};
    	assert program.merge(array1, array2) == new String[]{"good", "he", "is", "person"};
    	System.out.println("hakuna matata!!");
//    	String[] result = program.merge(array1, array2);
//    	for(int i = 0; i < result.length; i++){
//    		System.out.print(result[i] + " ");
//    	}
	}
}

File: Linkedin-Intership-Interview/R1Q2.py
# give two sorted array of strings, get a merged sorted array combining the two

def merge( array1, array2 ):
    m = len(array1)
    n = len(array2)
    i = 0
    j = 0
    resultArray = []
    while ( (i<m) and (j<n) ):
        if (array1[i] < array2[j]):
            resultArray.append( array1[i] )
            i += 1
        else:
            resultArray.append( array2[j] )
            j += 1

    if ( i < m ):
        resultArray.extend( array1[i:] )
    elif( j < n):
        resultArray.extend( array2[j:] )

    return resultArray

def merge2( array1, array2 ): # not needed!
    m = len(array1)  
    k = 0
    i = 0 # start
    j = m # start of array2
    resultArray = array1 + array2
    n = len( resultArray ) 
    while ( (i<m) and (j<n) ):
        if ( resultArray[j] < array1[i] ):
            resultArray[k] = resultArray[j]
            j += 1; k += 1
        else:
            resultArray[k] = array1[i]
            i += 1; k += 1

    while ( i < m ):
        resultArray[k] = array1[i]
        i += 1; k += 1
        
    return resultArray

def main():
    array1 = ["good", "person"] 
    array2 = ["he", "is"]
    assert merge(array1, array2) == ["good", "he", "is", "person"]
    print "\nAll asserts PASSED!!, Yaaaaaay!!\n"

if __name__ == "__main__":
    main()